datadog_checks_base/datadog_checks/base/utils/metadata/core.py

```diff
@@ -0,0 +1,138 @@
+            self.submit_raw(name, value)
+
+    def transform_version(self, version, options):
+        """Transforms a version like ``1.2.3-rc.4+5`` to its constituent parts. In all cases,
+        the metadata names ``version.raw`` and ``version.scheme`` will be sent.
+
+        If a ``scheme`` is defined then it will be looked up from our known schemes. If no
+        scheme is defined then it will default to semver.
+
+        The scheme may be set to ``regex`` in which case a ``pattern`` must also be defined. Any matching named
+        subgroups will then be sent as ``version.<GROUP_NAME>``. In this case, the check name will be used as
+        the value of ``version.scheme`` unless ``final_scheme`` is also set, which will take precedence.
+        """
+        scheme, version_parts = parse_version(version, options)
+        if scheme == 'regex':
+            scheme = options.get('final_scheme', self.check_name)
+
+        data = {'version.{}'.format(part_name): part_value for part_name, part_value in iteritems(version_parts)}
+        data['version.raw'] = version
+        data['version.scheme'] = scheme
+
+        return data
+
+    def transform_config(self, config, options):
+        """This transforms a ``dict`` of arbitrary user configuration. A ``section`` must be defined indicating
+        what the configuration represents e.g. ``init_config``.
+
+        The metadata name submitted will become ``config.<section>``.
+
+        The value will be a JSON ``str`` with the root being an array. There will be one map element for every
+        allowed field. Every map may have 2 entries:
+
+        1. ``is_set`` - a boolean indicating whether or not the field exists
+        2. ``value`` - the value of the field. this is only set if the field exists and the value is a
+                       primitive type (``None`` | ``bool`` | ``float`` | ``int`` | ``str``)
+
+        The allowed fields are derived from the optional ``whitelist`` and ``blacklist``. By default, nothing
+        will be sent.
+
+        User configuration can override defaults allowing complete, granular control of metadata submissions. In
+        any section, one may set ``metadata_whitelist`` and/or ``metadata_blacklist`` which will override their
+        keyword argument counterparts. In following our standard, blacklists take precedence over whitelists.
+        """
+        section = options.get('section')
+        if section is None:
+            raise ValueError('The `section` option is required')
+
+        # Although we define the default fields to send in code i.e. the default whitelist, there
+        # may be cases where a subclass (for example of OpenMetricsBaseCheck) would want to ignore
+        # just a few fields, hence for convenience we have the ability to also pass a blacklist.
+        whitelist = config.get('metadata_whitelist', options.get('whitelist', []))
+        blacklist = config.get('metadata_blacklist', options.get('blacklist', []))
```

Comment:
I feel that we should try to do something more secure by default. Otherwise we need each check to implement that logic, right? I feel that we can have at least a default blacklist that can be overridden. If we could reuse the credentials clear that the agent is using for its logs that'd be ideal.